src/cmap/conn/mod.rs

```diff
@@ -228,26 +212,130 @@ impl Connection {
-                strong_pool_ref.check_in(std::mem::replace(self, Self::null()));
-            } else if let Some(ref handler) = self.handler {
-                handler.handle_connection_closed_event(
-                    self.closed_event(ConnectionClosedReason::PoolClosed),
-                );
+                let dropped_connection_state = self.take();
+                RUNTIME.execute(async move {
+                    strong_pool_ref.check_in(dropped_connection_state.into()).await;
+                });
+            } else {
+                self.close(ConnectionClosedReason::PoolClosed);
             }
         }
     }
 }
+
+/// Options used for constructing a `Connection`.
+#[derive(Derivative)]
+#[derivative(Debug)]
+#[derive(Clone)]
+pub(crate) struct ConnectionOptions {
+    pub(crate) connect_timeout: Option<Duration>,
+    pub(crate) tls_options: Option<TlsOptions>,
+    #[derivative(Debug = "ignore")]
+    pub(crate) event_handler: Option<Arc<dyn CmapEventHandler>>,
+}
+
+impl From<ConnectionPoolOptions> for ConnectionOptions {
+    fn from(pool_options: ConnectionPoolOptions) -> Self {
+        Self {
+            connect_timeout: pool_options.connect_timeout,
+            tls_options: pool_options.tls_options,
+            event_handler: pool_options.event_handler,
+        }
+    }
+}
+
+/// Struct encapsulating the state of a connection that has been dropped.
+///
+/// Because `Drop::drop` cannot be async, we package the internal state of a dropped connection into this
+/// struct, and move it into an async task that will attempt to check the reconstructed connection back into the pool.
+///
+/// If the async runtime has been dropped, that task will not execute, and this state will be dropped. From there,
+/// we simply emit a connection closed event and do not attempt to reconstruct the `Connection`.
+#[derive(Derivative)]
+#[derivative(Debug)]
+struct DroppedConnectionState {
+    pub(super) id: u32,
+    pub(super) address: StreamAddress,
+    pub(super) generation: u32,
+    stream: Stream,
+    #[derivative(Debug = "ignore")]
+    handler: Option<Arc<dyn CmapEventHandler>>,
+    stream_description: Option<StreamDescription>,
+}
+
+impl Drop for DroppedConnectionState {
+    /// If this drop is called, that means the async runtime itself was dropped before the connection could be
+    /// checked back in to the pool. Instead of checking back into the pool again, we just close the connection
+    /// directly.
+    fn drop(&mut self) {
+        if let Some(ref handler) = self.handler {
+            handler.handle_connection_closed_event(ConnectionClosedEvent {
+                address: self.address.clone(),
+                connection_id: self.id,
+                reason: ConnectionClosedReason::PoolClosed,
+            });
+        }
+    }
+}
+
+impl From<DroppedConnectionState> for Connection {
+    fn from(mut state: DroppedConnectionState) -> Self {
+        Self {
```

Comment:
This is pretty similar to `Connection::take`; is there a way to share the code between them?

Reply:
Ah, good point